src/source/PeerConnection/PeerConnection.c

```diff
@@ -474,7 +474,66 @@ STATUS generateJSONSafeString(PCHAR pDst, UINT32 len)
     return retStatus;
 }
 
-STATUS createPeerConnection(PRtcConfiguration pConfiguration, PRtcPeerConnection *ppPeerConnection)
+STATUS rtcpReportsCallback(UINT32 timerId, UINT64 currentTime, UINT64 customData)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL ready       = FALSE;
+    UINT64 ntpTime, rtpTime, delay;
+    UINT32 packetCount, octetCount, packetLen, allocSize, ssrc;
+    PBYTE rawPacket                       = NULL;
+    PKvsPeerConnection pKvsPeerConnection = NULL;
+
+    PKvsRtpTransceiver pKvsRtpTransceiver = (PKvsRtpTransceiver) customData;
+    CHK(pKvsRtpTransceiver != NULL && pKvsRtpTransceiver->pJitterBuffer != NULL && pKvsRtpTransceiver->pKvsPeerConnection != NULL, STATUS_NULL_ARG);
+    pKvsPeerConnection = pKvsRtpTransceiver->pKvsPeerConnection;
+
+    ssrc = pKvsRtpTransceiver->sender.ssrc;
+    DLOGD("rtcpReportsCallback %" PRIu64 " ssrc: %u rtxssrc: %u", currentTime, ssrc, pKvsRtpTransceiver->sender.rtxSsrc);
+
+    // check if ice agent is connected, reschedule in 200msec if not
+    ready = pKvsPeerConnection->pSrtpSession != NULL &&
+        currentTime - pKvsRtpTransceiver->sender.firstFrameWallClockTime >= 2500 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+    if (!ready) {
+        DLOGD("sender report no frames sent %u", ssrc);
+    } else {
+        // create rtcp sender report packet
+        // https://tools.ietf.org/html/rfc3550#section-6.4.1
+        ntpTime = convertTimestampToNTP(currentTime);
+        rtpTime = pKvsRtpTransceiver->sender.rtpTimeOffset +
+            convertTimestampToRTP(pKvsRtpTransceiver->pJitterBuffer->clockRate, currentTime - pKvsRtpTransceiver->sender.firstFrameWallClockTime);
+        packetCount = pKvsRtpTransceiver->sender.outboundRtpStreamStats.sentRtpStreamStats.packetsSent;
+        octetCount  = pKvsRtpTransceiver->sender.outboundRtpStreamStats.sentRtpStreamStats.bytesSent;
+        DLOGD("sender report %u %" PRIu64 " %" PRIu64 " : %u packets %u bytes", ssrc, ntpTime, rtpTime, packetCount, octetCount);
+        packetLen = RTCP_PACKET_HEADER_LEN + 24;
+        allocSize = packetLen + SRTP_AUTH_TAG_OVERHEAD;
+        CHK(NULL != (rawPacket = (PBYTE) MEMALLOC(allocSize)), STATUS_NOT_ENOUGH_MEMORY);
+        rawPacket[0]                       = RTCP_PACKET_VERSION_VAL << 6;
+        rawPacket[RTCP_PACKET_TYPE_OFFSET] = RTCP_PACKET_TYPE_SENDER_REPORT;
+        putUnalignedInt16BigEndian(rawPacket + RTCP_PACKET_LEN_OFFSET,
+                                   (packetLen / RTCP_PACKET_LEN_WORD_SIZE) - 1); // The length of this RTCP packet in 32-bit words minus one
+        putUnalignedInt32BigEndian(rawPacket + 4, ssrc);
```

Comment:
Minor: Could we simply do SIZEOF(UINT32) instead of +4, etc? Normally, for packing, I use a current pointer which gets incremented by the size of the thing that I just copy. For example

put(pCurPtr, ssrc);
pCurPtr += SIZEOF(UINT32);
put(pCurPtr, ntpTime);
pCurPtr += SIZEOF(UINT64);
...

Reply:
i add explicit offsets to quickly debug what was written at which offset

Reply:
Really up to you. Some of these are minor comments/suggestions